File: ios/Runner/ChannelIOManager.swift
//
//  ChannelIOManager.swift
//  Runner
//
//  Created by 조희용 on 7/14/25.
//
// ChannelIOManager
import ChannelIOFront

class ChannelIOManager {
  static let CHANNEL_CHANNELIO = "com.example.SampleProject/channelIO"
  static let CHANNEL_BADGE_EVENT = "com.example.SampleProject/channelIO/badge"
  
  static let shared: ChannelIOManager = ChannelIOManager()
  
  private var eventSink: FlutterEventSink?
  
  func boot(bootConfig: [String:Any?]?, result: @escaping FlutterResult) {
    ChannelIO.delegate = self
    
    let config: BootConfig = BootConfig()
    config.pluginKey = bootConfig?["pluginKey"] as? String ?? ""
    config.memberId = bootConfig?["memberId"] as? String
    ChannelIO.boot(with: config) { bootStatus, user in
      result(bootStatus == .success ? "SUCCESS" : "ERROR")
    }
  }
  
  func sleep() {
    ChannelIO.sleep()
  }
  
  func shutdown() {
    ChannelIO.shutdown()
  }
  
  func showChannelButton() {
    ChannelIO.showChannelButton()
  }
  
  func showMessenger() {
    ChannelIO.showMessenger()
  }
  
  func track(eventName: String?) {
    guard let eventName = eventName else { return }
    ChannelIO.track(eventName: eventName)
  }
  
  func setBadgeListener(_ eventSink: FlutterEventSink?) {
    self.eventSink = eventSink
  }
  
  func openStoredPushNotification() {
    if ChannelIO.hasStoredPushNotification() {
      ChannelIO.openStoredPushNotification()
    }
  }

    func updateProfile(_ profile: [String:Any?]?) {
    guard let profile else { return }
    
    ChannelIO.updateUser(with: profile)
  }
}

extension ChannelIOManager: ChannelPluginDelegate {
  func onBadgeChanged(count: Int) {
    guard let eventSink = self.eventSink else { return }
    eventSink(count)
  }
}

File: android/app/src/main/kotlin/com/example/examproject1/ChannelIOManager.kt
package com.example.examproject1

import android.app.Activity
import android.util.Log
import com.tekartik.sqflite.Constant.TAG
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodChannel
import com.zoyi.channel.plugin.android.ChannelIO
import com.zoyi.channel.plugin.android.open.config.BootConfig
import com.zoyi.channel.plugin.android.open.enumerate.BootStatus
import com.zoyi.channel.plugin.android.open.listener.ChannelPluginListener
import com.zoyi.channel.plugin.android.open.model.PopupData
import com.zoyi.channel.plugin.android.open.model.User
import io.channel.com.google.gson.Gson

// ChannelIOManager
class ChannelIOManager : ChannelPluginListener {

    private var eventSink: EventChannel.EventSink? = null

    fun boot(bootConfig: HashMap<String, Any?>?, result: MethodChannel.Result) {
        val bootConfig = BootConfig.create("yourKey")
        ChannelIO.setListener(this)
        val config: BootConfig? = BootConfig.fromJson(Gson().toJson(bootConfig))
        ChannelIO.boot(config) { bootStatus: BootStatus, user: User? ->
            Log.e(TAG, "bootStatus: $bootStatus")
            result.success(bootStatus)
        }

    }

    fun sleep() {
        ChannelIO.sleep()
    }

    fun shutdown() {
        ChannelIO.shutdown()
    }

    fun showChannelButton() {
        ChannelIO.showChannelButton()
    }

    fun showMessenger(activity: Activity) {
        ChannelIO.showMessenger(activity)
    }

    fun track(eventName: String?) {
        eventName?.also { ChannelIO.track(it) }
    }

    fun setBadgeListener(eventSink: EventChannel.EventSink?) {
        this.eventSink = eventSink
    }

    override fun onShowMessenger() {
        TODO("Not yet implemented")
    }

    override fun onHideMessenger() {
        TODO("Not yet implemented")
    }

    override fun onChatCreated(p0: String?) {
        TODO("Not yet implemented")
    }

    override fun onBadgeChanged(count: Int) {
        this.eventSink?.success(count)
    }

    override fun onBadgeChanged(p0: Int, p1: Int) {
        TODO("Not yet implemented")
    }

    override fun onFollowUpChanged(p0: MutableMap<String, String>?) {
        TODO("Not yet implemented")
    }

    override fun onUrlClicked(p0: String?): Boolean {
        TODO("Not yet implemented")
    }

    override fun onPushNotificationClicked(p0: String?): Boolean {
        TODO("Not yet implemented")
    }

    override fun onPopupDataReceived(p0: PopupData?) {
        TODO("Not yet implemented")
    }

}